presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<PlanNode>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode apply, RewriteContext<PlanNode> context)
+        {
+            ApplyNode rewrittenApply = (ApplyNode) context.defaultRewrite(apply, context.get());
+            if (!rewrittenApply.getCorrelation().isEmpty()) {
+                Optional<ValuesNode> values = PlanNodeSearcher.searchFrom(apply.getSubquery())
+                        .skipOnlyWhen(ProjectNode.class::isInstance)
+                        .where(ValuesNode.class::isInstance)
+                        .findSingle();
+
+                if (values.isPresent() && isNoColumnSingleRowValues(values.get())) {
+                    List<ProjectNode> subqueryProjections = PlanNodeSearcher.searchFrom(apply.getSubquery())
+                            .where(ProjectNode.class::isInstance)
+                            .findAll();
+
+                    if (subqueryProjections.size() == 0) {
+                        return rewrittenApply.getInput();
+                    }
+                    else if (subqueryProjections.size() == 1) {
+                        return projectNode(
+                                projectNode(rewrittenApply.getInput(), subqueryProjections.get(0).getAssignments()),
+                                rewrittenApply.getSubqueryAssignments());
+                    }
+                }
+            }
+            return rewrittenApply;
+        }
+
+        private ProjectNode projectNode(PlanNode source, Map<Symbol, Expression> assignments)
+        {
+            Map<Symbol, Expression> newAssignments = new HashMap<>();
+            newAssignments.putAll(toAssignments(source.getOutputSymbols()));
+            appendAssignments(newAssignments, assignments);
+            return new ProjectNode(idAllocator.getNextId(), source, newAssignments);
+        }
+
+        private static void appendAssignments(Map<Symbol, Expression> assignments, Map<Symbol, Expression> toAdd)
+        {
+            for (Map.Entry<Symbol, Expression> assignment : toAdd.entrySet()) {
+                Symbol key = assignment.getKey();
+                Expression value = assignment.getValue();
+                if (assignments.containsKey(key)) {
+                    checkState(value.equals(assignments.get(key)));
+                }
+                else {
+                    if (value instanceof InPredicate) {
+                        InPredicate inPredicate = (InPredicate) value;
+                        value = new ComparisonExpression(ComparisonExpressionType.EQUAL, inPredicate.getValue(), inPredicate.getValueList());
+                    }
+                    assignments.put(assignment.getKey(), value);
```

Comment:
there might be other types of expressions that are invalid in projection context, e.g: `QuantifiedComparisonExpression`. I know that those are handled by `TransformQuantifiedComparisonApplyToScalarApply`, which is called before this optimizer, but I don't think you should rely on this assumption. Instead it would be better not to fire this optimizer when there are some unsupported `subqueryAssignments`.